### DISCUSSION TITLE
question related to .net7 with AOT

### DISCUSSION THREAD

**Discussion by zaafar** [amerkoleci/Vortice.Windows#379] [General]
When you were converting this project to .net 7 with AOT. did you encounter the following error? how did u fix it?

it's happening on the following line.

```
using var dxgiFactory = device.QueryInterface<IDXGIDevice>().GetParent<IDXGIAdapter>().GetParent<IDXGIFactory>();
```
![image](https://user-images.githubusercontent.com/5079646/212450846-4233d82c-a8ca-41b1-a8d2-89e91fd3f4f0.png)

**Comment by amerkoleci** [comment]
SharpGen had some trimming issues and some stuff has been done recently: https://github.com/SharpGenTools/SharpGenTools/commits/main

**Reply by zaafar** [reply]
Thank you, i will look into it.

**Reply by zaafar** [reply]
after doing a bit of light reading what I understand is following:

- SharpGen haven't been updated in last 6 months
- You are already using the latest version
- There is a PR out 6 months ago and that haven't been merged it

So, basically chances of this getting fixed is basically same as dead person walking. :)
Is there a workaround I can use until then?

**Reply by amerkoleci** [reply]
I'm one of the SharpGen maintainer as well

**Reply by amerkoleci** [reply]
Meanwhile you can use https://github.com/amerkoleci/Vortice.Win32 which are AOT friendly

**Reply by zaafar** [reply]
> I'm one of the SharpGen maintainer as well

oh, nice, so there is hope! :)

> Meanwhile you can use https://github.com/amerkoleci/Vortice.Win32 which are AOT friendly

thank you, let me check that out.

**Reply by zaafar** [reply]
holy shit! that PR just got merged.

**Comment by zaafar** [comment]
@amerkoleci i just tested the latest release, .net7 AOT is working! you are the best, thank you!!!!!

**Reply by amerkoleci** [reply]
Glad to hear that, spread the word! :)

**Comment by dmirmilshteyn** [comment]
Has this stopped working with .net 8? I'm trying AOT/trimming with the latest version and am getting a similar error (just with different classes like `IDXGIFactory2`). After compiling with trimming enabled, it looks like the constructors are being trimmed out (confirmed by looking at the compiled IL) even though the annotation is set on the SharpGenTools side.

Looking at the actual code, it seems there are lots of warnings related to this:
```
[IL2091] 'T' generic argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicConstructors' in 'SharpGen.Runtime.MarshallingHelpers.FromPointer<T>(nint)'. The generic parameter 'T' of 'Vortice.DXGI.DXGI.CreateDXGIFactory1<T>(out T)' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.
```

Am I just setting up trimming incorrectly, or do these annotations need to be added everywhere? (I'm willing to do all this and open a PR if you confirm this is the case)
